# Advent of Code 2023 - December 1st

### Day 1: Trebuchet?!
Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.

You've been doing this long enough to know that to restore snow operations, you need to check all fifty stars by December 25th.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

You try to ask why they can't just use a weather machine ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") and hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a trebuchet ("please hold still, we need to strap you in").

As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been amended by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.

The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover. On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:
```
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
```
In this example, the calibration values of these four lines are 12, 38, 15, and 77. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

In [79]:
f = open("./input1.txt")
contents = split(read(f, String))
close(f)

In [80]:
contents

1000-element Vector{SubString{String}}:
 "qzjggk1one"
 "two2seven7"
 "vszthreetwo6threethree4two3"
 "zcsvvlslqvfive11chhzmdjdgz8vbgldl"
 "jjsmsksvbr77cfdrdseven1zd"
 "947lkkgznlhxseventwo"
 "8twohprsxmz47"
 "five11eight1"
 "seventhree1eightztszfourfivesix"
 "cbcvd9"
 "hg91fourhhpvkrxbn"
 "1mqbdkqxnine817four"
 "btr4seventhree7hbkkg"
 ⋮
 "sixrzmqrkkd5fqqf"
 "sixkxtpcjmrmvvqjdgfttd6"
 "threehpbsevenffnqgdjcftjkdjhhk7dvzmkmqthreefflb"
 "6npp7fivebghnnxnqjxhhxfdqpsixfour"
 "ffour87fqrvqxqlqrrk"
 "6ninerzksfbszmqnvgmtqonesixjzsf"
 "sevenszl3lkqqfhlrdmkxgggvmkb"
 "5c1"
 "vrxbfckfz7qlfrnggsbdfive6glhk62three"
 "7eightfiveeightrdnhfnqeight5"
 "5threethreerdlsfdsvggzbonecxqzqdvxcqtqbg1"
 "gttsix2567"

In [81]:
function firstandlast(ln)
    firstindex = findfirst(isdigit, ln)
    lastindex = findlast(isdigit, ln)
    parse(Int, string(ln[firstindex], ln[lastindex]))
end

sum(firstandlast.(contents))

55172

Result:55172

#### Part Two
Your calculation isn't quite right. It looks like some of the digits are actually spelled out with letters: one, two, three, four, five, six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real first and last digit on each line. For example:

```
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
```
In this example, the calibration values are 29, 83, 13, 24, 42, 14, and 76. Adding these together produces 281.

What is the sum of all of the calibration values?

In [82]:
function digitize(name)
    if name == "one"
        "1"
    elseif name == "two"
        "2"
    elseif name == "three"
        "3"
    elseif name == "four"
        "4"
    elseif name == "five"
        "5"
    elseif name == "six"
        "6"
    elseif name == "seven"
        "7"
    elseif name == "eight"
        "8"
    elseif name == "nine"
        "9"
    else
        name
    end
end

function firstlastwithnames(ln)
    i = digitize.([match.match for match in eachmatch(r"1|2|3|4|5|6|7|8|9|one|two|three|four|five|six|seven|eight|nine", ln, overlap=true)])
    parse(Int, string(i[begin], i[end]))
end

sum(firstlastwithnames.(contents))

54925

Result:54925